fix(docs): complete apidocs for migrations

bjoerge opened the pull request:
### Description

<!--
What changes are introduced?
Why are these changes introduced?
What issue(s) does this solve? (with link, if possible)
-->

### What to review

<!--
What steps should the reviewer take in order to review?
What parts/flows of the application/packages/tooling is affected?
-->

### Testing

<!--
Did you add sufficient testing for this change?
If not, please explain how you tested this change and why it was not
possible/practical for writing an automated test
-->

### Notes for release

<!--
A description of the change(s) that should be used in the release notes.
-->

kmelve commented on packages/@sanity/migrate/src/mutations/creators.ts:
```diff
@@ -33,7 +33,7 @@ export function create<Doc extends Optional<SanityDocument, '_id'>>(
  * @param id - The id of the document to be patched.
```
```suggestion
 * @param id - The ID of the document to be patched.
```

kmelve commented on packages/@sanity/migrate/src/mutations/creators.ts:
```diff
@@ -98,6 +98,6 @@ export function delete_(id: string): DeleteMutation {
 
 /**
- * Alias for delete_
+ * Alias for delete
```
```suggestion
 * Alias for deleting a document by ID
```
A bit of a stub. Not sure if the suggestion is accurate?

bjoerge replied:
`delete` is a reserved keyword in JS, so you can't import it like this:
`import {delete} from 'sanity/migrate // syntax error`

So it's exported as both `delete_` with `del` as an alias, so you can do 
`import {del} from 'sanity/migrate // ok` or `import {delete_} from 'sanity/migrate // ok`

kmelve commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
````diff
@@ -41,6 +52,11 @@ export const setIfMissing = <const T>(value: T): SetIfMissingOp<T> => ({
+ *
+ * ### Example
+ * ```ts
+ * const unsetAnyValue = unset()
````
```suggestion
 * const unsetAnyValue = unset()
 * const unsetArrayItemByIndex = unset('list[1]')
 * const unsetArrayItemByFieldValue = unset('list[_key=="foo"]')
```

Adding some more advanced examples (I _think_ they're accurate?).

bjoerge replied:
No, unset doesn't take a path, so it'd be: `at('list[_key=="foo"]', unset())`

kmelve commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -50,6 +66,12 @@ export const unset = (): UnsetOp => ({type: 'unset'})
  * @returns An `inc` operation.
```
```suggestion
 * @returns An incrementation operation for numeric values
```

kmelve commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -62,6 +84,12 @@ export const inc = <const N extends number = 1>(amount: N = 1 as N): IncOp<N> =>
  * @returns A `dec` operation.
```
```suggestion
 * @returns An decrementation operation for numeric values
```

bjoerge replied:
Isn't it more commonly referred to as "a decrement" operation? (Realized too late it should probably also be "An increment operation" above, if so)

kmelve commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -88,6 +116,13 @@ export const diffMatchPatch = (value: string): DiffMatchPatchOp => ({
  * @returns An `insert` operation.
```
```suggestion
 * @returns An `insert` operation for adding values to arrays
```

kmelve commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -112,6 +147,13 @@ export function insert<
  * @returns An `insert` operation at the end of the array.
```
```suggestion
 * @returns An `insert` operation for adding a value to the end of an array.
```

kmelve commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -123,6 +165,13 @@ export function append<const Items extends AnyArray<unknown>>(items: Items | Arr
  * @returns An `insert` operation at the beginning of the array.
```
```suggestion
 * @returns An `insert` operation for adding a value to the start of an array.
```

kmelve commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -55,42 +86,44 @@ export type ReplaceOp<
-}
 
+/**
+ * Represents a diffMatchPatch operation that can be applied to a string
```
```suggestion
 * Represents a diff-match-patch operation that can be applied to a string
 * {@link https://code.google.com/p/google-diff-match-patch/}
```

kmelve commented on packages/@sanity/migrate/src/mutations/transaction.ts:
```diff
@@ -5,7 +5,17 @@ export interface Transaction {
+ * Wraps a set of mutations in a transaction.
+ * Note: use with caution. Transactions cannot be optimized and will be submitted one-by-one, which means they will make
+ * your migration run slower and produce more API requests.
+ * @param transactionId - The transaction id. This is optional and should usually be omitted, as it will be auto-generated by the server if not provided.
```
```suggestion
 * @param transactionId - The transaction ID. This is optional and should usually be omitted, as it will be auto-generated by the server if not provided.
```

kmelve commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -2,12 +2,18 @@ import type {Path} from '@sanity/types'
   | readonly [NodePatch, ...NodePatch[]]
 
+/**
+ * A sanity document
```
```suggestion
 * A Sanity Content Lake document 
```

kmelve commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -2,12 +2,18 @@ import type {Path} from '@sanity/types'
+/**
+ * A sanity document
+ */
 export type SanityDocument = {
```
Don't we have this type elsewhere in the Sanity code base? Won't it always have `_updatedAt` and `_createdAt` as well?

bjoerge replied:
Somehow missed this comment.

We do have it defined in quite a few places already, but there's not a huge of value in re-using it IMO as long as the definitions are structurally compatible. This type is mostly used with `create({_type: 'foo'})` et. al., so making `_updatedAt`, `_createdAt` etc. mandatory would require you to specify those when creating new documents from a migration.

However, I agree these fields could/should be defined as optional strings.

kmelve commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -16,45 +22,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a create-mutation that can be applied to a Sanity document.
```
```suggestion
 * Represents a mutation that can be applied to create a new document in the Sanity Content Lake.
```

kmelve commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -16,45 +22,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a createIfNotExists-mutation that can be applied to a Sanity document if it does not exist.
```
```suggestion
 * Represents a mutation that can create a new document in the Sanity Content Lake if its ID does not exist.
```

kmelve commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -16,45 +22,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a createOrReplace mutation that can be applied to a Sanity document.
```
```suggestion
 * Represents a mutation that can create or replace a document in the Sanity Content Lake given its ID.
```

kmelve commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -16,45 +22,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a delete mutation that can be applied to a Sanity document.
```
```suggestion
 * Represents a mutation that can delete a document in the Sanity Content Lake.
```

kmelve commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -16,45 +22,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a patch mutation that can be applied to a Sanity document.
```
```suggestion
 * Represents a patch mutation that can change a value for a document in the Sanity Content Lake.
```

kmelve commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -16,45 +22,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a mutation that can be applied to a Sanity document.
```
```suggestion
 * Represents a mutation that can be applied to a document in the Sanity Content Lake.
```

kmelve commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -16,45 +22,69 @@ export type SanityDocument = {
+ * Options for a patch operation.
+ */
 export type PatchOptions = {
   ifRevision?: string
```
Can we add docs for `ifRevision` ? Not sure where to put that.

bjoerge replied:
Yep! Can probably refer to this section: https://www.sanity.io/docs/http-mutations#26600a871378

`patch(doc._id, [at('foo', set('bar'))], {ifRevision: doc._rev})`

binoy14 replied:
maybe? 
```suggestion
  /**
  * {@link https://www.sanity.io/docs/http-mutations#26600a871378}
  */
  ifRevision?: string
```

kmelve reviewed (commented):
Great start! Left several suggestions. Do read carefully before approving as I'm not 100% confident I got it all right.

binoy14 commented on packages/@sanity/migrate/src/defineMigration.ts:
```diff
@@ -1,5 +1,29 @@
+ *
+ * @see https://www.sanity.io/docs/schema-and-content-migrations#af2be129ccd6
+
+ * ### Basic usage
```
need to use `@example` so it picks it up correctly

```suggestion
 * @example Basic usage
```

binoy14 commented on packages/@sanity/migrate/src/defineMigration.ts:
```diff
@@ -1,5 +1,29 @@
+ * Helper function for defining a Sanity content migration. This function does not do anything on its own;
+ * it exists to check that your schema definition is correct, and help autocompletion in your IDE.
+ *
+ * @see https://www.sanity.io/docs/schema-and-content-migrations#af2be129ccd6
```
`@see` is not support need to use `@link`

```suggestion
 * {@link https://www.sanity.io/docs/schema-and-content-migrations#af2be129ccd6}
```

binoy14 commented on packages/@sanity/migrate/src/defineMigration.ts:
```diff
@@ -1,5 +1,29 @@
 import {type Migration} from './types'
 
+/**
+ * Helper function for defining a Sanity content migration. This function does not do anything on its own;
```
missing `@public` 
```suggestion
 * @public
 *
 * Helper function for defining a Sanity content migration. This function does not do anything on its own;
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -22,6 +22,12 @@ import {
  * {@link https://www.sanity.io/docs/http-patches#6TPENSW3}
  * @beta
+ *
+ * ### Examples
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -31,6 +37,11 @@ export const set = <const T>(value: T): SetOp<T> => ({type: 'set', value})
  * @returns A `setIfMissing` operation.
  * {@link https://www.sanity.io/docs/http-patches#A80781bT}
  * @beta
+ * ### Examples
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -42,15 +53,26 @@ export const setIfMissing = <const T>(value: T): SetIfMissingOp<T> => ({
  * {@link https://www.sanity.io/docs/http-patches#xRtBjp8o}
  * @beta
+ *
+ * ### Example
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -42,15 +53,26 @@ export const setIfMissing = <const T>(value: T): SetIfMissingOp<T> => ({
  * {@link https://www.sanity.io/docs/http-patches#vIT8WWQo}
  * @beta
+ *
+ * ### Example
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -63,6 +85,12 @@ export const inc = <const N extends number = 1>(amount: N = 1 as N): IncOp<N> =>
  * {@link https://www.sanity.io/docs/http-patches#vIT8WWQo}
  * @beta
+ *
+ * ### Example
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -74,7 +102,7 @@ export const dec = <const N extends number = 1>(amount: N = 1 as N): DecOp<N> =>
  * @returns A `diffMatchPatch` operation.
  * {@link https://www.sanity.io/docs/http-patches#aTbJhlAJ}
- * @beta
+ *
```
```suggestion
 * @public
```

binoy14 replied:
or beta

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -86,9 +114,16 @@ export const diffMatchPatch = (value: string): DiffMatchPatchOp => ({
  * {@link https://www.sanity.io/docs/http-patches#febxf6Fk}
  * @beta
+ *
+ * ### Example
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -113,6 +148,13 @@ export function insert<
  * {@link https://www.sanity.io/docs/http-patches#Cw4vhD88}
  * @beta
+ *
+ * ### Example
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -124,6 +166,13 @@ export function append<const Items extends AnyArray<unknown>>(items: Items | Arr
  * {@link https://www.sanity.io/docs/http-patches#refAUsf0}
  * @beta
+ *
+ * ### Example
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -135,6 +184,12 @@ export function prepend<const Items extends AnyArray<unknown>>(items: Items | Ar
  * @returns An `insert` operation before the provided index or reference item.
  * {@link https://www.sanity.io/docs/http-patches#0SQmPlb6}
+ *
+ * ### Example
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -135,6 +184,12 @@ export function prepend<const Items extends AnyArray<unknown>>(items: Items | Ar
  * @param indexOrReferenceItem - The index or reference item to insert before.
  * @returns An `insert` operation before the provided index or reference item.
  * {@link https://www.sanity.io/docs/http-patches#0SQmPlb6}
+ *
```
```suggestion
 * @public
 *
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -150,6 +205,12 @@ export function insertBefore<
  * {@link https://www.sanity.io/docs/http-patches#0SQmPlb6}
  * @beta
+ *
+ * ### Example
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -169,6 +230,13 @@ export const insertAfter = <
  * {@link https://www.sanity.io/docs/http-patches#xRtBjp8o}
  * @beta
+ *
+ * ### Example
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/creators.ts:
```diff
@@ -186,6 +254,13 @@ export function truncate(startIndex: number, endIndex?: number): TruncateOp {
  * {@link https://www.sanity.io/docs/http-patches#GnVSwcPa}
  * @beta
+ *
+ * ### Example
```
```suggestion
 * @example
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -2,36 +2,60 @@ import {type KeyedSegment} from '@sanity/types'
 import {type AnyArray} from '../typeUtils'
 
+/**
+ * Represents an indexed segment in a document.
```
```suggestion
 * @public
 *
 * Represents an indexed segment in a document.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -2,36 +2,60 @@ import {type KeyedSegment} from '@sanity/types'
 export type {IndexedSegment, KeyedSegment}
 
+/**
+ * Represents a set-operation that can be applied to any value
```
```suggestion
 * @public
 *
 * Represents a set-operation that can be applied to any value
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -2,36 +2,60 @@ import {type KeyedSegment} from '@sanity/types'
 }
 
+/**
+ * Represents an unset operation that can be applied to any value
```
```suggestion
 * @public
 *
 * Represents an unset operation that can be applied to any value
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -2,36 +2,60 @@ import {type KeyedSegment} from '@sanity/types'
 }
 
+/**
+ * Represents a setIfMissing operation that can be applied to any value
```
```suggestion
 * @public
 *
 * Represents a setIfMissing operation that can be applied to any value
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -2,36 +2,60 @@ import {type KeyedSegment} from '@sanity/types'
 }
 
+/**
+ * Represents a increment-operation that can be applied to a number
```
```suggestion
 * @public
 *
 * Represents a increment-operation that can be applied to a number
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -2,36 +2,60 @@ import {type KeyedSegment} from '@sanity/types'
 }
 
+/**
+ * Represents a decrement-operation that can be applied to a number
```
```suggestion
 * @public
 *
 * Represents a decrement-operation that can be applied to a number
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -2,36 +2,60 @@ import {type KeyedSegment} from '@sanity/types'
 }
 
+/**
+ * Represents a relative position in a document.
```
```suggestion
 * @public
 *
 * Represents a relative position in a document.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -2,36 +2,60 @@ import {type KeyedSegment} from '@sanity/types'
 export type RelativePosition = 'before' | 'after'
 
+/**
+ * Represents an insert-operation that can be applied to an array
```
```suggestion
 * @public
 *
 * Represents an insert-operation that can be applied to an array
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -43,11 +67,18 @@ export type InsertOp<
 }
 
+/**
+ * Represents a truncate-operation that can be applied to an array
```
```suggestion
 * @public
 *
 * Represents a truncate-operation that can be applied to an array
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -43,11 +67,18 @@ export type InsertOp<
 }
+
+/**
+ * Represents a replace-operation that can be applied to an array
```
```suggestion
 * @public
 *
 * Represents a replace-operation that can be applied to an array
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -56,42 +87,44 @@ export type ReplaceOp<
-}
 
+/**
+ * Represents a diffMatchPatch operation that can be applied to a string
```
```suggestion
 * @public
 *
 * Represents a diffMatchPatch operation that can be applied to a string
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -56,42 +87,44 @@ export type ReplaceOp<
 
-export type Operation = PrimitiveOp | ArrayOp | ObjectOp
+/**
+ * Represents an operation that can be applied to values of all types
```
```suggestion
 * @public
 *
 * Represents an operation that can be applied to values of all types
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -56,42 +87,44 @@ export type ReplaceOp<
+export type Operation = PrimitiveOp | ArrayOp
 
+/**
+ * Represents an operation that can be applied to values of all types
```
```suggestion
 * @public
 *
 * Represents an operation that can be applied to values of all types
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -56,42 +87,44 @@ export type ReplaceOp<
 export type AnyOp = SetOp<unknown> | SetIfMissingOp<unknown> | UnsetOp
+
+/**
+ * Represents an operation that can be applied to a number
```
```suggestion
 * @public
 *
 * Represents an operation that can be applied to a number
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -56,42 +87,44 @@ export type ReplaceOp<
 export type NumberOp = IncOp<number> | DecOp<number>
+
+/**
+ * Represents an operation that can be applied to a string
```
```suggestion
 * @public
 *
 * Represents an operation that can be applied to a string
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -56,42 +87,44 @@ export type ReplaceOp<
-export type ObjectOp = AssignOp | UnassignOp
+
+/**
+ * Represents ann operation that can be applied to an array
```
```suggestion
 * @public
 *
 * Represents ann operation that can be applied to an array
```

binoy14 commented on packages/@sanity/migrate/src/mutations/operations/types.ts:
```diff
@@ -56,42 +87,44 @@ export type ReplaceOp<
   | TruncateOp
 
+/**
+ * Represents an operation that can be applied to any primitive value
```
```suggestion
 * @public
 *
 * Represents an operation that can be applied to any primitive value
```

binoy14 commented on packages/@sanity/migrate/src/mutations/transaction.ts:
```diff
@@ -5,7 +5,17 @@ export interface Transaction {
-export function transaction(id: string, mutations: Mutation[]): Transaction
+
+/**
+ * Wraps a set of mutations in a transaction.
```
```suggestion
 * @public
 *
 * Wraps a set of mutations in a transaction.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/transaction.ts:
```diff
@@ -5,7 +5,17 @@ export interface Transaction {
+ * @param transactionId - The transaction id. This is optional and should usually be omitted, as it will be auto-generated by the server if not provided.
+ * @param mutations - The mutations to include in the transaction.
+ *
+ * @see https://www.sanity.io/docs/http-mutations#afccc1b9ef78
```
```suggestion
 * {@link https://www.sanity.io/docs/http-mutations#afccc1b9ef78}
```

binoy14 commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -3,12 +3,18 @@ import {type Path} from '@sanity/types'
 import {type Optional} from './typeUtils'
 
+/**
+ * A list of {@link NodePatch} objects.
```
```suggestion
 * @public
 *
 * A list of {@link NodePatch} objects.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -3,12 +3,18 @@ import {type Path} from '@sanity/types'
   | readonly [NodePatch, ...NodePatch[]]
 
+/**
+ * A sanity document
```
```suggestion
 * @public
 *
 * A sanity document
```

binoy14 commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -17,45 +23,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a create-mutation that can be applied to a Sanity document.
```
```suggestion
 * @public
 *
 * Represents a create-mutation that can be applied to a Sanity document.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -17,45 +23,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a createIfNotExists-mutation that can be applied to a Sanity document if it does not exist.
```
```suggestion
 * @public
 *
 * Represents a createIfNotExists-mutation that can be applied to a Sanity document if it does not exist.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -17,45 +23,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a mutation that can create or replace a document in the Sanity Content Lake given its ID.
```
```suggestion
 * @public
 *
 * Represents a mutation that can create or replace a document in the Sanity Content Lake given its ID.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -17,45 +23,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a mutation that can delete a document in the Sanity Content Lake.
```
```suggestion
 * @public
 *
 * Represents a mutation that can delete a document in the Sanity Content Lake.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -17,45 +23,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a patch mutation that can change a value for a document in the Sanity Content Lake.
```
```suggestion
 * @public
 *
 * Represents a patch mutation that can change a value for a document in the Sanity Content Lake.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -17,45 +23,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Represents a mutation that can be applied to a document in the Sanity Content Lake.
```
```suggestion
 * @public
 *
 * Represents a mutation that can be applied to a document in the Sanity Content Lake.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -17,45 +23,69 @@ export type SanityDocument = {
   | PatchMutation
 
+/**
+ * A NodePatch represents a single operation that can be applied at a node at a specific path in a Sanity document.
```
```suggestion
 * A NodePatch represents a single operation that can be applied at a node at a specific path in a Sanity document.
```
```suggestion
 * @public
 *
 * A NodePatch represents a single operation that can be applied at a node at a specific path in a Sanity document.
```

binoy14 commented on packages/@sanity/migrate/src/mutations/types.ts:
```diff
@@ -17,45 +23,69 @@ export type SanityDocument = {
 }
 
+/**
+ * Options for a patch operation.
```
```suggestion
 * @public
 *
 * Options for a patch operation.
```

binoy14 commented on packages/@sanity/migrate/src/runner/utils/batchMutations.ts:
```diff
@@ -14,7 +14,6 @@ function isTransactionPayload(payload: any): payload is TransactionPayload {
  * @param mutations - Async iterable of either single values or arrays of values
  * @param maxBatchSize - Max batch size in bytes
- * Todo: add support for transaction ids too
  */
```
missing public
```suggestion
 * @public
 *
 */
```

binoy14 commented on packages/@sanity/migrate/src/types.ts:
```diff
@@ -13,31 +13,63 @@ export type AsyncIterableMigration = (
 
+/**
+ * Main interface for a content migration definition
+ * @see https://www.sanity.io/docs/schema-and-content-migrations#af2be129ccd6
```
```suggestion
 * {@link https://www.sanity.io/docs/schema-and-content-migrations#af2be129ccd6}
```

binoy14 commented on packages/@sanity/migrate/src/types.ts:
```diff
@@ -13,31 +13,63 @@ export type AsyncIterableMigration = (
 ) => AsyncGenerator<Mutation | Transaction | (Mutation | Transaction)[]>
 
+/**
+ * Main interface for a content migration definition
```
```suggestion
 * @public
 *
 * Main interface for a content migration definition
```

binoy14 commented on packages/@sanity/migrate/src/types.ts:
```diff
@@ -13,31 +13,63 @@ export type AsyncIterableMigration = (
 }
 
+/**
+ * Migration context. This will be passed to both async iterable migrations and node migration helper functions
```
```suggestion
 * @public
 *
 * Migration context. This will be passed to both async iterable migrations and node migration helper functions
```

binoy14 commented on packages/@sanity/migrate/src/types.ts:
```diff
@@ -13,31 +13,63 @@ export type AsyncIterableMigration = (
+}
+
+/**
+ * Interface for `Migration['migrate']`. Either a NodeMigration or an AsyncIterableMigration
```
```suggestion
 * @public
 *
 * Interface for `Migration['migrate']`. Either a NodeMigration or an AsyncIterableMigration
```

binoy14 commented on packages/@sanity/migrate/src/types.ts:
```diff
@@ -13,31 +13,63 @@ export type AsyncIterableMigration = (
 
+/**
+ * Migration progress, only used internally (for now)
+ * @internal
```
```suggestion
 * @internal
 * @hidden
```

binoy14 commented on packages/@sanity/migrate/src/types.ts:
```diff
@@ -48,14 +80,10 @@ export type MigrationProgress = {
-
+/**
+ * API configuration for the migration runner
+ * @internal
```
```suggestion
 * @internal
 * @hidden
```

binoy14 commented on packages/@sanity/migrate/src/types.ts:
```diff
@@ -64,10 +92,18 @@ export interface APIConfig {
 
+/**
+ * API configuration for exports
+ * @internal
```
```suggestion
 * @internal
 * @hidden
```

binoy14 commented on packages/@sanity/migrate/src/types.ts:
```diff
@@ -64,10 +92,18 @@ export interface APIConfig {
 }
 
+/**
+ * Possible return values from a migration helper that runs on a document as a whole
```
```suggestion
 * @public
 *
 * Possible return values from a migration helper that runs on a document as a whole
```

binoy14 commented on packages/@sanity/migrate/src/types.ts:
```diff
@@ -76,9 +112,20 @@ export type DocumentMigrationReturnValue =
   | RawMutation[]
 
+/**
+ * Possible return values from a migration helper that runs on nodes within a document
```
```suggestion
 * @public
 *
 * Possible return values from a migration helper that runs on nodes within a document
```

binoy14 commented on packages/@sanity/migrate/src/types.ts:
```diff
@@ -76,9 +112,20 @@ export type DocumentMigrationReturnValue =
 export type NodeMigrationReturnValue = DocumentMigrationReturnValue | Operation | Operation[]
 
+/**
+ * Node migration helper functions. As the migration is processing a document, it will visit each node in the document, depth-first, call the appropriate helper function for the node type and collect any mutations returned from it.
```
```suggestion
 * @public
 *
 * Node migration helper functions. As the migration is processing a document, it will visit each node in the document, depth-first, call the appropriate helper function for the node type and collect any mutations returned from it.
```

binoy14 reviewed (changes requested):
Adding some missing tags and suggestions for not supported tags. I'll leave the review of copy to @kmelve 

P.S: not sure if you want to show beta items in the docs but if not they would need `@hidden` tag

binoy14 commented:
there is a draft https://github.com/sanity-io/sanity/pull/5664/ if you want to preview it locally copy the changes from there and `pnpm build && pnpm etl sanity && pnpm tsdoc:dev` should do the trick

bjoerge commented:
Thanks for all the great feedback @kmelve and @binoy14! 🙏🏼 

I've addressed all of your suggestions/comments now (I hope!) and pushed in a separate commit. Please re-review and make sure I haven't overlooked anything (GitHub's UI keeps confusing me 😵‍💫)

binoy14 commented on packages/@sanity/migrate/tsdoc.json:
I believe this won't be needed because we re-export this from the sanity package and once we get #5664 merged into all these docs will be exposed from the sanity package. It would become redundant  to expose both but maybe there is a good reason to have it?

bjoerge replied:
I needed it for `@hidden` to work with tsdoc compilation (IIRC). Let's see if we can remove it after https://github.com/sanity-io/sanity/pull/5664 lands then?

binoy14 replied:
ah right, you are correct it's needed for `@hidden` I forgot that it still isolated

binoy14 reviewed (approved):
LGTM thanks! small question on the `tsdoc.json` file besides that looks good

kmelve reviewed (approved):
Looks great!
